### PR TITLE
SPU LLVM Improvements

### PR DESCRIPTION
**Pull request by Nekotekina**
Fixes PS1 Classics with SPU LLVM.

### CODE REVIEW

**Comment by Xcedf** [comment]
MGS1 section in MGS4 now works on SPU LLVM

**Comment by Yahfz** [comment]
Crysis 3 now also works with spu llvm. 
image

**Comment by CookiePLMonster** [comment]
PS1 Classics fixed indeed.

image

File: rpcs3/Emu/Cell/SPURecompiler.cpp [hunk 1]
```diff
@@ -465,6 +465,29 @@ void spu_cache::initialize()
...
+				sha1_update(&ctx, reinterpret_cast<const u8*>(func.data.data()), func.data.size() * 4);
+				sha1_finish(&ctx, output);
+				std::memcpy(&hash_start, output, sizeof(hash_start));
+			}
+
+			// Check hash against allowed bounds
+			const bool inverse_bounds = g_cfg.core.spu_llvm_lower_bound > g_cfg.core.spu_llvm_upper_bound;
+
+			if (!inverse_bounds && (hash_start < g_cfg.core.spu_llvm_lower_bound || hash_start > g_cfg.core.spu_llvm_upper_bound) ||
+				inverse_bounds && (hash_start < g_cfg.core.spu_llvm_lower_bound && hash_start > g_cfg.core.spu_llvm_upper_bound))
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/SPURecompiler.cpp:483]: inverse_bounds check is redundant

File: rpcs3/Emu/Cell/SPURecompiler.cpp [hunk 2]
```diff
@@ -8833,7 +8871,15 @@ struct spu_fast : public spu_recompiler_base
 		// Install pointer carefully
 		const bool added = !add_loc->compiled && add_loc->compiled.compare_and_swap_test(nullptr, fn);
 
-		if (added)
+		// Check hash against allowed bounds
+		const bool inverse_bounds = g_cfg.core.spu_llvm_lower_bound > g_cfg.core.spu_llvm_upper_bound;
+
+		if (!inverse_bounds && (m_hash_start < g_cfg.core.spu_llvm_lower_bound || m_hash_start > g_cfg.core.spu_llvm_upper_bound) ||
+			inverse_bounds && (m_hash_start < g_cfg.core.spu_llvm_lower_bound && m_hash_start > g_cfg.core.spu_llvm_upper_bound))
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/SPURecompiler.cpp:8878]: same

File: rpcs3/Emu/Cell/SPURecompiler.cpp (hunk 1 above)
Comment (Megamouse) [reply rpcs3/Emu/Cell/SPURecompiler.cpp:483]: ups. sorry

**Pull request merged**